internal/gitserver/client.go

```diff
@@ -614,6 +631,163 @@ var deadlineExceededCounter = promauto.NewCounter(prometheus.CounterOpts{
 	Help: "Times that Client.sendExec() returned context.DeadlineExceeded",
 })
 
+// BatchLog invokes the given callback with the `git log` output for a batch of repository
+// and commit pairs. If the invoked callback returns a non-nil error, the operation will begin
+// to abort processing further results.
+func (c *ClientImplementor) BatchLog(ctx context.Context, opts BatchLogOptions, callback BatchLogCallback) (err error) {
+	ctx, endObservation := c.operations.batchLog.With(ctx, &err, observation.Args{LogFields: []log.Field{
+		log.Int("opts.numRepoCommits", len(opts.RepoCommits)),
+		log.String("opts.Format", opts.Format),
+	}})
+	defer endObservation(1, observation.Args{})
+
+	// Make a request to a signle gitserver shard and feed the results to the user-supplied
+	// callback. Invoked multiple times (and concurrently) in the function defined in the
+	// following.
+	performLogRequestToShard := func(addr string, repoCommits []api.RepoCommit) (err error) {
+		var numProcessed int
+		repoNames := repoNamesFromRepoCommits(repoCommits)
+
+		ctx, logger, endObservation := c.operations.batchLogSingle.WithAndLogger(ctx, &err, observation.Args{
+			LogFields: []log.Field{
+				log.String("addr", addr),
+				log.Int("numRepos", len(repoNames)),
+				log.Int("numRepoCommits", len(repoCommits)),
+			},
+		})
+		defer func() {
+			endObservation(1, observation.Args{
+				LogFields: []log.Field{
+					log.Int("numProcessed", numProcessed),
+				},
+			})
+		}()
+
+		uri := "http://" + addr + "/batch-log"
+		repoName := api.RepoName(strings.Join(repoNames, ",")) // only used to label spans
+
+		request := protocol.BatchLogRequest{
+			RepoCommits: repoCommits,
+			Format:      opts.Format,
+		}
+
+		var buf bytes.Buffer
+		if err := json.NewEncoder(&buf).Encode(request); err != nil {
+			return err
+		}
+
+		resp, err := c.do(ctx, repoName, "POST", uri, buf.Bytes())
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		logger.Log(log.Int("resp.StatusCode", resp.StatusCode))
+
+		if resp.StatusCode != http.StatusOK {
+			body, _ := io.ReadAll(io.LimitReader(resp.Body, 200))
+			return errors.Newf("http status %d: %s", resp.StatusCode, body)
+		}
+
+		var response protocol.BatchLogResponse
+		if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+			return err
+		}
+		logger.Log(log.Int("numResults", len(response.Results)))
+
+		for _, result := range response.Results {
+			var err error
+			if result.CommandError != "" {
+				err = errors.Newf(result.CommandError)
```

Comment:
```suggestion
				err = errors.New(result.CommandError)
```

To avoid (mis-)interpreting any formatting specifiers accidentally present in `result.CommandError`.